codalab/lib/upload_manager.py

```diff
@@ -258,3 +261,210 @@ def cleanup_existing_contents(self, bundle):
         bundle_update = {'data_hash': None, 'metadata': {'data_size': 0}}
         self._bundle_model.update_bundle(bundle, bundle_update)
         self._bundle_model.update_user_disk_used(bundle.owner_id)
+
+    def get_bundle_sas_token(self, path, **kwargs):
+        """
+        Get SAS token with write permission. Used for bypass server uploading.
+        """
+        return (
+            parse_linked_bundle_url(path)
+            .bundle_path_sas_url(permission='rw', **kwargs)
+            .split('?')[-1]
+        )
+
+    def get_index_sas_token(self, path, **kwargs):
+        """
+        Get SAS token of the index file with read and write permission. Used for uploading.
+        """
+        return (
+            parse_linked_bundle_url(path)
+            .index_path_sas_url(permission='rw', **kwargs)
+            .split('?')[-1]
+        )
+
+    def get_bundle_signed_url(self, path, **kwargs):
+        """
+        Get signed url for the bundle path
+        """
+        return parse_linked_bundle_url(path).bundle_path_signed_url(**kwargs)
+
+    def get_bundle_index_url(self, path, **kwargs):
+        return parse_linked_bundle_url(path).index_path_signed_url(**kwargs)
+
+
+class ClientUploadManager(object):
+    """
+    Upload Manager for CLI client. Handle file upload for CLI client.
+    """
+
+    def __init__(self, json_api_client, stdout, stderr):
+        self._client = json_api_client
+        self.stdout = stdout
+        self.stderr = stderr
+
+    def upload_to_bundle_store(
+        self,
+        bundle: Dict,
+        packed_source: Dict,
+        use_azure_blob_beta: bool,
+        destination_bundle_store=None,
+    ):
+
+        # By pass server upload:
+        # 1. If the user specify `-a`, upload to Azure blob storage
+        # 2. If the user specify `--store` and blob storage is on Azure or GCS
+        bypass_server = True if use_azure_blob_beta else False
+        bundle_store_uuid = None
+        if destination_bundle_store is not None and destination_bundle_store != '':
+            storage_info = self._client.fetch_one(
+                'bundle_stores',
+                params={
+                    'name': destination_bundle_store,
+                    'include': ['uuid', 'storage_type', 'url'],
+                },
+            )
+            bundle_store_uuid = storage_info['uuid']
+            if storage_info['storage_type'] in (
+                StorageType.AZURE_BLOB_STORAGE.value,
+                StorageType.GCS_STORAGE.value,
+            ):
+                bypass_server = True
+
+        if bypass_server:
+            # Mimic the rest server behavior
```

Comment:
> Initialize
> 
> `Uploader(is_client=True, client=self._client)`
> 
> In `Uploader. _update_and_get_bundle_location`, if `is_client` is True, then it should call `add_bundle_location`, save the SAS URLs, then set them to the appropriate env variables right when uploading happens.

I reuse the function write_fileobj() function and keep the `_update_and_get_bundle_location` the same.  It's not easy to reuse the Upload._update_and_get_bundle_location function because it uses a lot self._bundle_store/_bundle_model. If we reuse it, we basically need to have an if statement and rewrite all the parts. I feel it's more clear to handle this in ClientUploadManger.

Reply:
Sounds good